### PR TITLE
Move ParsedMethods into structuring phase

### PR DESCRIPTION
**Pull request by BenFordTytherington**
All methods are now collected in structuring phase, as `ParsedMethod`, and then associated with `StructuredQObjects`, which have a `HashMap<Ident, &ParsedMethod>` which stores the `rust_unqualified` name as a key, and the method as the value, so that they can be looked up.

### CODE REVIEW

File: crates/cxx-qt-gen/src/parser/cxxqtdata.rs [hunk 1]
```diff
@@ -21,6 +21,9 @@ use syn::{
 };
 use syn::{ItemMacro, Meta};
 
+// PROTOTYPING ONLY
+use crate::naming::Name;
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/parser/cxxqtdata.rs:25]: think clippy is raging about this :-)

File: crates/cxx-qt-gen/src/generator/cpp/qobject.rs [hunk 2]
```diff
@@ -175,7 +177,7 @@ impl GeneratedCppQObject {
             let (initializer, mut blocks) = threading::generate(&qobject_idents)?;
             generated.blocks.append(&mut blocks);
             class_initializers.push(initializer);
-        // If this type has locking enabled then add generation
+            // If this type has locking enabled then add generation
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/generator/cpp/qobject.rs:180]: that looks like a fun formatting change, maybe move this comment down into the else if block below as it's supposed to be for that block

```
} else if qobject.locking {
    // If this type has locking enabled then add generation
```

File: crates/cxx-qt-gen/src/generator/structuring/mod.rs [hunk 3]
```diff
@@ -29,15 +31,30 @@ impl<'a> Structures<'a> {
     /// Create a new `Structures` object from the given `ParsedCxxQtData`
     /// Returns an error, if any references could not be resolved.
     pub fn new(cxxqtdata: &'a ParsedCxxQtData) -> Result<Self> {
+        let methods_found: Vec<_> = cxxqtdata
+            .methods
+            .iter()
+            .map(|method| (method.name.clone(), method.qobject_ident.clone()))
+            .collect();
+        for method_pair in methods_found {
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/generator/structuring/mod.rs:39]: i know it's just tracing but remember you can destructure :-)

`for (key, value) in methods_found {`

File: crates/cxx-qt-gen/src/generator/structuring/mod.rs [hunk 4]
```diff
@@ -54,6 +71,30 @@ impl<'a> Structures<'a> {
...
         }
 
+        for method in &cxxqtdata.methods {
+            println!(
+                "Looking for QObject match for method: {:?}",
+                method.name.clone()
+            );
+            if let Some(qobject_ident) = qobjects.iter_mut().find(|qobject| {
+                qobject.declaration.name.rust_unqualified() == &method.qobject_ident
+            }) {
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/generator/structuring/mod.rs:81]: This looks quite intense to be doing for every method we iterate over all qobjects?

Does this point that we should store the qobjects above as a hash instead by their rust declaration name?

```
let mut qobjects: HashMap<Ident, StructuredQObject>
```

Or should there be a cache of the index or a mut ref (might be very tricky with lifetimes). Eg `HashMap<Ident, usize>` or `HashMap<Ident, &mut StructuredQobject>` so we can get from the qobject ident -> structured qobject easily?

File: crates/cxx-qt-gen/src/parser/cxxqtdata.rs [hunk 5]
```diff
@@ -204,31 +213,38 @@ impl ParsedCxxQtData {
...
+                    let parsed_signal_method_self = ParsedSignal::parse(foreign_fn, safe_call)?;
+                    self.signals.push(parsed_signal_method_self);
 
                     self.with_qobject(&parsed_signal_method.qobject_ident)?
                         .signals
                         .push(parsed_signal_method);
-                // Test if the function is an inheritance method
-                //
-                // Note that we need to test for qsignal first as qsignals have their own inherit meaning
+                    // Test if the function is an inheritance method
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/parser/cxxqtdata.rs:224]: seems to be in lots of places :-) Is this jetbrains formatting subtly differently ? or has Rust 1.80 changed things :thinking: 

As it's creating noise unrelated to your change, either should be in it's own change if that is a cargo fmt change or we need to figure out why your IDE is subtly different :sweat_smile:

File: crates/cxx-qt-gen/src/parser/cxxqtdata.rs (hunk 1 above)
Comment (BenFordTytherington) [reply crates/cxx-qt-gen/src/parser/cxxqtdata.rs]: Done

File: crates/cxx-qt-gen/src/generator/cpp/signal.rs [hunk 6]
```diff
@@ -521,7 +521,7 @@ mod tests {
         let mut type_names = TypeNames::default();
         type_names.mock_insert("ObjRust", None, None, None);
         let qobject_name = type_names.lookup(&signal.qobject_ident).unwrap();
-        let generated = generate_cpp_signal(&signal, &qobject_name, &type_names).unwrap();
+        let generated = generate_cpp_signal(&signal, qobject_name, &type_names).unwrap();
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/generator/cpp/signal.rs:524]: where is this change coming from as the signature hasn't changed ? Seems like a clippy style fix ?

File: crates/cxx-qt-gen/src/generator/rust/qobject.rs [hunk 7]
```diff
@@ -23,6 +23,7 @@ use quote::quote;
 use syn::{Ident, Result};
 
 impl GeneratedRustFragment {
+    // Might need to be refactored to use a StructuredQObject instead (confirm with Leon)
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/generator/rust/qobject.rs:26]: seems likely :-)

File: crates/cxx-qt-gen/src/generator/cpp/signal.rs (hunk 6 above)
Comment (BenFordTytherington) [reply crates/cxx-qt-gen/src/generator/cpp/signal.rs:524]: yeah, was "immediately dereferencing" warning, so assuming qobject_name was already a reference, so function got &&T and just instantly coerced to &T, so it wasn't needed

File: crates/cxx-qt-gen/src/generator/structuring/mod.rs [hunk 8]
```diff
@@ -54,6 +51,113 @@ impl<'a> Structures<'a> {
...
+            if let Some(qobject) = qobjects
+                .iter_mut()
+                .find(|qobject| qobject.has_qobject_name(&method.qobject_ident))
+            {
+                qobject.methods.push(method);
+            } else {
+                return Err(Error::new_spanned(
+                    &method.qobject_ident,
+                    format!("Unknown QObject: {:?}", &method.qobject_ident),
+                ));
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/generator/structuring/mod.rs:65]: Wonder if this is more readable then these big if let Some blocks ?

```rust
            let mut qobject = qobjects
                .iter_mut()
                .find(|qobject| qobject.has_qobject_name(qobject_ident))
                .ok_or_else(|| {
                    Error::new_spanned(qobject_ident, format!("Unknown QObject: {qobject_ident}"))
                })?;
            qobject.methods.push(method);
```

Comment (BenFordTytherington) [reply crates/cxx-qt-gen/src/generator/structuring/mod.rs:65]: Done

File: crates/cxx-qt-gen/src/parser/cxxqtdata.rs (hunk 5 above)
Comment (BenFordTytherington) [reply crates/cxx-qt-gen/src/parser/cxxqtdata.rs:222]: Done

File: crates/cxx-qt-gen/src/generator/rust/qobject.rs (hunk 7 above)
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/generator/rust/qobject.rs:26]: Yup, assuming the next one is working on the generator phase

File: crates/cxx-qt-gen/src/generator/cpp/qobject.rs (hunk 2 above)
Comment (BenFordTytherington) [reply crates/cxx-qt-gen/src/generator/cpp/qobject.rs:179]: Removed

File: crates/cxx-qt-gen/src/parser/cxxqtdata.rs (hunk 1 above)
Comment (BenFordTytherington) [reply crates/cxx-qt-gen/src/parser/cxxqtdata.rs]: Removed

File: crates/cxx-qt-gen/src/generator/cpp/signal.rs (hunk 6 above)
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/generator/cpp/signal.rs:524]: Even running with `--all-targets` doesn't help fun :-)

**Review by ahayzen-kdab — APPROVED** [review]
LGTM, lets go with this as a first step of adding structuring support for methods then generation phase refactor is another PR.

**Pull request merged**